base/fillarrays.jl

```diff
@@ -0,0 +1,134 @@
+ 7  7  7
+ 7  7  7
+
+julia> Fill{Float64, 1, Tuple{UnitRange{Int64}}}(7., (1:2,))
+2-element Fill{Float64,1,Tuple{UnitRange{Int64}}} with indices 1:2:
+ 7.0
+ 7.0
+```
+
+For more `Fill` functionality and efficient routines, use the FillArrays.jl package.
+"""
+struct Fill{T, N, Axes} <: AbstractFill{T, N, Axes}
+    value::T
+    axes::Axes
+
+    Fill{T,N,Axes}(x::T, sz::Axes) where Axes<:Tuple{Vararg{AbstractUnitRange,N}} where {T, N} =
+        new{T,N,Axes}(x,sz)
+    Fill{T,0,Tuple{}}(x::T, sz::Tuple{}) where T = new{T,0,Tuple{}}(x, sz)
+end
+
+Fill{T,N,Axes}(x, sz::Axes) where Axes<:Tuple{Vararg{AbstractUnitRange,N}} where {T, N} =
+    Fill{T,N,Axes}(convert(T, x)::T, sz)
+
+Fill{T,0}(x::T, ::Tuple{}) where T = Fill{T,0,Tuple{}}(x, ()) # ambiguity fix
+
+@inline Fill{T, N}(x::T, sz::Axes) where Axes<:Tuple{Vararg{AbstractUnitRange,N}} where {T, N} =
+    Fill{T,N,Axes}(x, sz)
+@inline Fill{T, N}(x, sz::Axes) where Axes<:Tuple{Vararg{AbstractUnitRange,N}} where {T, N} =
+    Fill{T,N}(convert(T, x)::T, sz)
+
+@inline Fill{T, N}(x, sz::SZ) where SZ<:Tuple{Vararg{Integer,N}} where {T, N} =
+    Fill{T,N}(x, Base.OneTo.(sz))
+@inline Fill{T, N}(x, sz::Vararg{Integer, N}) where {T, N} = Fill{T,N}(convert(T, x)::T, sz)
+
+
+@inline Fill{T}(x, sz::Vararg{<:Integer,N}) where {T, N} = Fill{T, N}(x, sz)
+@inline Fill{T}(x, sz::Tuple{Vararg{<:Any,N}}) where {T, N} = Fill{T, N}(x, sz)
+""" `Fill(x, dims...)` construct lazy version of `fill(x, dims...)` """
+@inline Fill(x::T, sz::Vararg{<:Integer,N}) where {T, N}  = Fill{T, N}(x, sz)
+""" `Fill(x, dims)` construct lazy version of `fill(x, dims)` """
+@inline Fill(x::T, sz::Tuple{Vararg{<:Any,N}}) where {T, N}  = Fill{T, N}(x, sz)
+
+# We restrict to  when T is specified to avoid ambiguity with a Fill of a Fill
+@inline Fill{T}(F::Fill{T}) where T = F
+@inline Fill{T,N}(F::Fill{T,N}) where {T,N} = F
+@inline Fill{T,N,Axes}(F::Fill{T,N,Axes}) where {T,N,Axes} = F
+
+@inline axes(F::Fill) = F.axes
+@inline size(F::Fill) = length.(F.axes)
+
+AbstractArray{T}(F::Fill{T}) where T = F
+AbstractArray{T,N}(F::Fill{T,N}) where {T,N} = F
+AbstractArray{T}(F::Fill{V,N}) where {T,V,N} = Fill{T}(convert(T, F.value)::T, F.axes)
+AbstractArray{T,N}(F::Fill{V,N}) where {T,V,N} = Fill{T}(convert(T, F.value)::T, F.axes)
+
+convert(::Type{AbstractArray{T}}, F::Fill{T}) where T = F
+convert(::Type{AbstractArray{T,N}}, F::Fill{T,N}) where {T,N} = F
+convert(::Type{AbstractArray{T}}, F::Fill) where {T} = AbstractArray{T}(F)
+convert(::Type{AbstractArray{T,N}}, F::Fill) where {T,N} = AbstractArray{T,N}(F)
+convert(::Type{AbstractFill}, F::AbstractFill) = F
+convert(::Type{AbstractFill{T}}, F::AbstractFill) where T = convert(AbstractArray{T}, F)
+convert(::Type{AbstractFill{T,N}}, F::AbstractFill) where {T,N} = convert(AbstractArray{T,N}, F)
+
+copy(F::Fill) = Fill(F.value, F.axes)
+
+getindex(F::Fill{<:Any,0}) = getindex_value(F)
+
+@inline getindex_value(F::Fill) = F.value
```

Comment:
Do we need this?  Why not just define this in `getindex` and use `F[]` everywhere?

Reply:
That seems inconsistent to me unless you meant something else:
```julia
julia> F = fill(1,3)
3-element Vector{Int64}:
 1
 1
 1

julia> F[]
ERROR: BoundsError: attempt to access 3-element Vector{Int64} at index []
Stacktrace:
 [1] throw_boundserror(A::Vector{Int64}, I::Tuple{})
   @ Base ./abstractarray.jl:651
 [2] checkbounds
   @ ./abstractarray.jl:616 [inlined]
 [3] _getindex
   @ ./abstractarray.jl:1196 [inlined]
 [4] getindex(::Vector{Int64})
   @ Base ./abstractarray.jl:1170
 [5] top-level scope
   @ REPL[20]:1
```

Reply:
By `F[]` I just meant indexing in general.  Seems like an unnecessary indirection/complexity.

Reply:
It's to allow for unified code for `Fill`, `Zeros`, and `Ones`. Using indexing doesn't work here as one would have to add artificial indices.